File: euler.clj
(require '(clojure [string :as s]))

(defmacro euler [n body] `(defmethod solution ~n [n#] (delay ~body)))
(defmulti solution (fn [n] n))
(defmethod solution :default [n] (delay "UNSOLVED"))
(defn solve [ens]
  (doseq [n ens
        :let [start (System/nanoTime)
              sol @(solution n)
              stop (System/nanoTime)
              elapsed (float (/ (- stop start) 1000000000))]]
    (println (format "%03d: %s (%.3fs)" n sol elapsed))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn multiples-of [divisors]
  (filter
    (fn [n] (some #(zero? (mod n %)) divisors))
    (rest (range))))

(def fib-seq (lazy-cat [1 2]
  (map + fib-seq (rest fib-seq))))

(defn prime-factors [num]
  (loop [n num f 2 factors []]
    (cond
      (= n 1) (distinct factors)
      (zero? (mod n f)) (recur (quot n f) f (conj factors f))
      :else (recur n (inc f) factors))))

(defn palindrome? [n]
  (= (str n) (s/reverse (str n))))

(defn gcd [a b] (loop [a a b b] (if (zero? b) a (recur b (mod a b)))))
(defn lcm [a b] (/ (* a b) (gcd a b)))

(defn square [n] (* n n))
(def squares (lazy-seq (map square (rest (range)))))

(defn sum-of-squares [n] (reduce + (take n squares)))
(defn square-of-sum [n] (square (reduce + (range (inc n)))))

(defn nth-prime [n]
  (loop [p [2] i 3]
    (cond
      (= (count p) n) (last p)
      (some #(zero? (mod i %)) (take-while #(<= % (Math/sqrt i)) p)) (recur p (+ i 2))
      :else (recur (conj p i) (+ i 2)))))

(declare primes)

(defn prime? [n]
  (let [factors (take-while #(<= % (Math/sqrt n)) primes)]
    (not-any? #(zero? (mod n %)) factors)))

(def primes
  (lazy-cat [2]
    (filter prime? (take-nth 2 (iterate inc 3)))))

(defn readlines [filename] (s/split-lines (slurp filename)))

; partition then reduce over the partitions
(defn part-reduce [size reduce-fn nums]
  (let [parts (partition size 1 nums)]
    (map #(reduce reduce-fn %) parts)))

(defn transpose [m]
  (apply vector (apply map vector m)))

(defn diagonals [m]
  (let [n (count m)
        at (fn [i j] (get (get m i) j))]
    (for [diag (range (dec (* 2 n)))
          :let [startrow (max 0 (inc (- diag n)))]]
      (for [col (range startrow (inc (- diag startrow)))
            :let [row (- diag col)]]
        (at row col)))))

(defn diagonalsr [m]
  (-> m reverse vec diagonals))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

; 001: Find the sum of all the multiples of 3 or 5 below 1000.
(euler 1 (reduce + (take 1000 (multiples-of [3,5]))))

; 002: Find the sum of even-valued fibonacci numbers below 4M
(euler 2 (reduce + (filter even? (take-while #(< % 4000000) fib-seq))))

; 003: What is the largest prime factor of 600851475143?
(euler 3 (reduce max (prime-factors 600851475143)))

; 004: Find the largest palindrome made from the product of two 3-digit numbers.
(euler 4
  (apply max
    (filter palindrome?
      (for [i (range 100 999) j (range 100 999)] (* i j)))))

; 005: Find the smallest positive number evenly divisible by all of (1..20)
(euler 5 (reduce lcm (range 1 20)))

; 006: Find the difference of sum of squares and square of the sum of (1..100)
(euler 6
  (let [n 100] (- (square-of-sum n) (sum-of-squares n))))

; 007: What is the 10001st prime number?
(euler 7 (nth primes 10000))

; 008: Find the greatest product of five consecutive digits in a 1000-digit number.
(euler 8
  (let [n (map #(Integer. (str %))
               (.replaceAll (slurp "data/008") "[\r\n]+" ""))]
    (reduce max (part-reduce 5 * n))))

; 009: Find the product abc for the pythagorean triple where a + b + c = 1000.
(euler 9 (first
  (for [n (range 1 1000) m (range 1 n)
        :let [a (- (* n n) (* m m))
              b (* 2 n m)
              c (+ (* n n) (* m m))]
        :when (= 1000 (+ a b c))]
    (* a b c))))

; 010: Find the sum of all the primes below two million.
(euler 10 (reduce + (take-while #(< % 2000000) primes)))

; 011: Find the largest product of four adjacent numbers in a grid
(euler 11
  (let [toi #(Integer. %)
      break #(s/split % #"\s+")
      data (readlines "data/011")
      grid (mapv #(mapv toi (break %)) data)
      maxr #(reduce max (flatten %))
      adjmax (fn [m] (maxr (remove empty? (map #(part-reduce 4 * %) m))))]
    (reduce max (map adjmax
      [grid (transpose grid) (diagonals grid) (diagonalsr grid)]))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;
(let [args (seq *command-line-args*)]
  (solve (map #(Integer. %) (or args (range 1 11)))))
;;;;;;;;;;;;;;;;;;;;;;;;;;;

File: euler.hs
import Text.Printf
import Data.List
import Data.Maybe

-- returns true if f is a factor of n
divides :: Integer -> Integer -> Bool
divides f n = n `rem` f == 0

-- returns True if all of fs are factors of n
hasFactors :: [Integer] -> Integer -> Bool
hasFactors fs n = all (\f -> f n) (map divides fs)

-- a lazy list of fibonacci numbers
fibs :: [Integer]
fibs = 0 : 1 : zipWith (+) fibs (tail fibs)

-- returns the prime factors of n
primeFactors :: Integer -> [Integer]
primeFactors n = factorize 2 n where
  factorize _ 1 = []
  factorize d n
    | d * d > n = [n]
    | d `divides` n = d : factorize d (n `div` d)
    | otherwise = factorize (d + 1) n

-- returns primes
primes :: [Integer]
primes = 2 : filter (null . tail . primeFactors) [3,5..]

-- returns True if n is a palindrome number
isPalindrome :: Integer -> Bool
isPalindrome n = (show n) == (reverse $ show n)

-- returns pythagorean triples below
pythag :: [(Integer, Integer, Integer)]
pythag = [(a,b,c) | m <- [2..],
                    n <- [1..(m-1)],
                    let a = m^2 - n^2,
                    let b = 2*n*m,
                    let c = n^2 + m^2,
                    a^2 + b^2 == c^2]

solve :: Integer -> Integer
-- 001: Find the sum of all the multiples of 3 or 5 below 1000.
solve 1 = sum $ [n | n <- [1..999], 3 `divides` n || 5 `divides` n]
-- 002: Find the sum of even-valued fibonacci numbers below 4M
solve 2 = sum $ filter even $ takeWhile (< 4000000) fibs
-- 003: What is the largest prime factor of 600851475143?
solve 3 = last $ primeFactors 600851475143
-- 004: Find the largest palindrome made from the product of two 3-digit numbers.
solve 4 = maximum [ z | x <- [100..999], y <- [100..999], let z = x*y, isPalindrome z]
-- 005: Find the smallest positive number evenly divisible by all of (1..20)
solve 5 = foldl1 lcm [1..20]
-- 006: Find the difference of sum of squares and square of the sum of (1..100)
solve 6 = ((^2) $ sum [1..100]) - (sum [x^2 | x <- [1..100]])
-- 007: What is the 10001st prime number?
solve 7 = primes !! 10000
-- 008: Find the greatest product of five consecutive digits in a 1000-digit number.
-- 009: Find the product abc for the pythagorean triple where a + b + c = 1000.
solve 9 = let (a,b,c) = fromJust $ (find (\(a,b,c) -> a + b + c == 1000) pythag) in a*b*c
-- 010: Find the sum of all the primes below two million.
-- 011: Find the largest product of four adjacent numbers in a grid
-- 012: What is the first triangle number to have over 500 divisors?
-- 013: What are the first ten digits of the sum of 100 50-digit numbers?
-- 014: Find the starting number that produces the longest chain
solve n = -1

main = do
  mapM_ putStrLn [printf "%03d: %d" n (solve n) | n <- [1..15]]
